File: main/coin-change/coin_change.rb
#!/usr/bin/env ruby
$VERBOSE = true

def count_ways(coins, total)
  memo = {}

  solve = lambda do |index, subtot|
    return 1 if subtot.zero?
    return 0 if index == coins.size

    key = [index, subtot].freeze
    count = memo[key]
    return count if count

    count = 0
    value = coins[index]
    index += 1

    until subtot.negative?
      count += solve.call(index, subtot)
      subtot -= value
    end

    memo[key] = count
  end

  solve.call(0, total)
end

def read_record
  gets.split.map!(&:to_i)
end

total, = read_record # need n (total), but not m (record length)
coins = read_record
puts count_ways(coins, total)

File: main/coin-change/coin-change.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Carp qw(croak);

sub read_record {
    return split /\s+/msx, <>;
}

my ($total, $len) = read_record;
my @coins = read_record;
$len == scalar @coins or croak 'wrong record length';

my %memo; # "$index $subtotal" => $count

sub count_ways {
    my ($index, $subtotal) = @_;

    if ($subtotal == 0) { return 1 }
    if ($index == $len) { return 0 }

    my $key = "$index $subtotal";
    my $count = $memo{$key};

    if (!defined $count) {
        my $coin = $coins[$index];
        my $next_index = $index + 1;

        $count = 0;
        while ($subtotal >= 0) {
            $count += count_ways($next_index, $subtotal);
            $subtotal -= $coin;
        }

        $memo{$key} = $count;
    }

    return $count;
}

print count_ways(0, $total), "\n";
